### PR TITLE
feat: implement configurable key allowlist to pass keys through shield (Issue #64)

### LINKED ISSUE #64: feat: add configurable key allowlist to pass specific keys through the shield
## Summary

Add a configurable allowlist of keys and/or key combinations that will NOT be blocked when Cat Shield is active. This allows users to maintain access to specific keyboard shortcuts (like media controls, Spotlight, or other global hotkeys) while still protecting their work from cat interference.

## Problem Statement

Currently, Cat Shield blocks ALL keyboard input except the exit key combination. This is effective for cat protection but can be overly restrictive when users need:

- **Media controls**: Volume up/down, play/pause, next/previous track
- **System shortcuts**: Spotlight (`Cmd+Space`), Alfred/Raycast, Mission Control
- **Accessibility features**: VoiceOver shortcuts, zoom controls
- **App-specific global hotkeys**: Clipboard managers, window managers, etc.

## Proposed Solution

Add a new setting `allowed_keys` that accepts a list of key combinations using the same format as the exit key (e.g., `Cmd+Space`, `F11`, `Ctrl+Option+A`).

### Config File Format

```toml
exit_key = "Cmd+Option+U"
default_timer = "30m"
overlay_opacity = 0.6

# NEW: Keys that pass through the shield (not blocked)
allowed_keys = [
  "Cmd+Space",          # Spotlight
  "F10",                # Mute
  "F11",                # Volume down
  "F12",                # Volume up
  "Ctrl+Up",            # Mission Control
  "Ctrl+Down",          # App windows
]
```

### Behavior

1. **Exit key still exits**: The exit key combination continues to deactivate the shield
2. **Allowlist keys pass through**: Keys in the allowlist are allowed to reach the system/apps without blocking
3. **All other keys blocked**: Everything not in the allowlist or exit key is blocked as usual
4. **Empty allowlist = current behavior**: No keys pass through (backward compatible)

### Files to Modify

| File | Changes |
|------|---------|
| `src/config/types.rs` | Add `allowed_keys: Option<Vec<String>>` to `Config` struct |
| `src/platform/event_tap.rs` | Check allowlist before blocking in `event_tap_callback` |
| `src/input/exit_key.rs` | Add `AllowedKeys` struct/functions for parsing and matching |
| `src/ui/windows/settings.rs` | Add allowlist UI section |
| `src/ui/state.rs` | Add UI state pointers for allowlist components |

### Event Tap Callback Flow (Updated)

```
User presses key
    ↓
HID Event Tap captures
    ↓
Is it the exit key? ─→ Yes → Deactivate shield
    ↓ No
Is it in the allowlist? ─→ Yes → Return event (allow through)
    ↓ No
Return NULL → Block event
```

### Performance Considerations

- Use `HashSet<(i64, bool, bool, bool, bool)>` for O(1) lookup (keycode + modifiers tuple)
- Parse allowlist once at config load, not on every keystroke
- Store parsed allowlist in static atomic state like exit key

### Settings UI Design

Add a new section "Allowed Keys" between Overlay Opacity and the button row:

```
┌─────────────────────────────────────────────────┐
│  Allowed Keys                                    │
│  Keys that pass through the shield               │
│  ┌────────────────────────────────────────────┐ │
│  │ Cmd+Space                              [X] │ │
│  │ F11                                    [X] │ │
│  │ F12                                    [X] │ │
│  │ Ctrl+Up                                [X] │ │
│  └────────────────────────────────────────────┘ │
│  [+ Add Key] [Clear All]                         │
│                                                  │
│  Add: [____________] [Add]                       │
│  Format: Modifier+Key (e.g., Cmd+Space, F11)    │
│  ✓ Valid key combination                         │
└─────────────────────────────────────────────────┘
```

UI Components:
- Scrollable list showing current allowed keys
- Remove button (X) for each entry
- Text field to add new keys (reuses exit key validation)
- Real-time validation with green checkmark or red error
- "Clear All" to reset the list

### 1. Exit Key in Allowlist
If a user adds the exit key to the allowlist, it should still function as an exit key (not just pass through). The exit key check should happen first.

### 2. Modifier-Only Entries
Do not allow modifier-only entries (e.g., just "Cmd" or "Shift") as these are ambiguous and could interfere with modifier detection.

### 3. Function Keys on Modern Macs
Modern MacBooks map F1-F12 to media functions by default. Consider:
- Document that users need F-key (not media key) mode enabled, OR
- Add preset buttons: "Add media keys (F10-F12)" for common use case

### 4. Maximum Allowlist Size
Consider a reasonable maximum (e.g., 50 entries) to prevent performance issues and maintain the shield's effectiveness.

### 5. Duplicate Prevention
Silently ignore duplicates when adding to the list.

### 6. Case Handling
Key names should be case-insensitive (reuse existing exit key parsing logic).

## Acceptance Criteria

- [ ] Users can add/remove key combinations from allowlist in Settings
- [ ] Allowlist persists to `~/.config/catshield/config.toml`
- [ ] Keys in allowlist pass through when shield is active
- [ ] Exit key still functions normally (exits shield)
- [ ] Real-time validation when adding keys (same rules as exit key)
- [ ] Empty allowlist maintains current behavior (all keys blocked)
- [ ] Settings UI displays current allowlist with remove option
- [ ] Clear All button removes all allowlist entries
- [ ] Unit tests for allowlist parsing and matching
- [ ] Integration test for key passthrough behavior

## Future Enhancements (Separate Issues)

- **Preset groups** (e.g., "Media Keys", "System Shortcuts" one-click add) - will create follow-up issue
- Import/export allowlist configurations
- Per-session allowlist (different from saved config)
- Visual indicator when allowed key is pressed

## Related Code

- Exit key implementation: `src/input/exit_key.rs`
- Event tap callback: `src/platform/event_tap.rs:event_tap_callback()`
- Config system: `src/config/types.rs`

### PR DESCRIPTION
**Pull request by taearls**
This commit implements a comprehensive key allowlist feature that allows users
to specify keyboard shortcuts that won't be blocked by the shield overlay.

## Config & Data Model
- Added `allowed_keys: Option<Vec<String>>` field to Config struct
- Added comprehensive tests for config serialization/deserialization
- Config format: `allowed_keys = ["Cmd+Space", "F11", "F12"]`

## Core Logic
- Created `src/input/allowed_keys.rs` module for managing allowed keys
  - `AllowedKey` struct with parsing and matching logic
  - Global RwLock-protected state for thread-safe key storage
  - `parse_and_set_allowed_keys()` for bulk key validation
  - `is_key_allowed()` for event filtering
- Supports same format as exit keys (e.g., "Cmd+Space", "F11", "Ctrl+Option+A")
- Also supports simple keys without modifiers (e.g., "F11", "Space")

## Event Filtering
- Modified event tap callback in `src/platform/event_tap.rs`
- Checks if incoming key events match allowed keys before blocking
- Allowed keys pass through to the system, all others are blocked
- Maintains exit key functionality as highest priority

## Shield Integration
- Load allowed keys from config during shield activation
- Display configured keys in activation banner
- Clear allowed keys on shield deactivation
- Handle invalid keys gracefully with warning messages

## Settings UI
- Increased settings window height from 370 to 550 pixels
- Added "Allowed Keys" section with:
  - Scrollable text view showing current allowed keys (one per line)
  - Input field for adding new keys with placeholder hint
  - Real-time validation as user types
  - "Add" button to add keys to the list
  - "Clear All" button to remove all keys
  - Duplicate detection to prevent adding same key twice
- Created `AddKeyFieldDelegate` for real-time input validation
- Added action handlers: `addAllowedKey:` and `clearAllowedKeys:`
- Keys are added/removed in memory, persisted on "Save" click
- "Reset to Default" clears allowed keys list

## UI State Management
- Added atomic pointers to `src/ui/state.rs`:
  - `ALLOWED_KEYS_VIEW` - NSTextView displaying keys
  - `ALLOWED_KEYS_SCROLL` - NSScrollView wrapper
  - `ADD_KEY_FIELD` - Input field for new keys
  - `ADD_KEY_VALIDATION` - Validation message label
  - `ADD_KEY_FIELD_DELEGATE` - Real-time validation delegate
- Proper cleanup on window close

## Testing
- Added 6 new config tests for allowed_keys field
- Added 13 new unit tests in allowed_keys module covering:
  - Parsing with/without modifiers
  - Validation and error handling
  - Global state management
  - Event matching logic
- All existing tests continue to pass

## Use Cases
This feature enables:
- Media controls (F10-F12) while shield is active
- Spotlight (Cmd+Space) for quick app launching
- Mission Control and other system shortcuts
- Custom workflows with specific key combinations

Closes #64

## Summary by CodeRabbit

* **New Features**
  * Configurable key allowlist: define key combinations that pass through the shield; add/validate/clear keys via Settings.

* **Behavioral Changes**
  * Keyboard handling refined: KeyDown/KeyUp/modifier events respected; allowlisted keys pass through, others are blocked; activation loads configured keys, deactivation clears them.

* **UI**
  * Settings window gains Allowed Keys list, input field, real‑time validation, Add and Clear All controls.

* **Tests**
  * Added parsing, matching, concurrency, and config round‑trip tests.

* **Documentation**
  * Roadmap, changelog, and status counts updated.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/ui/windows/settings.rs [hunk 1]
```diff
@@ -1353,6 +1394,342 @@ fn setup_button_section(
...
+    settings::ADD_KEY_VALIDATION.store(
+        Retained::as_ptr(&validation_label) as *mut c_void,
+        Ordering::Release,
+    );
+    std::mem::forget(validation_label);
+
+    y_offset -= 30.0;
+
+    y_offset
+}
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/ui/windows/settings.rs]: Fixed in commit 855346c. All three compilation errors have been resolved:
1. `create_label` calls now have all required arguments
2. `setDelegate` now uses `get_or_create_add_key_delegate()` helper function (matching pattern used by other delegates)
3. Both `create_label` calls are properly formatted

File: src/ui/windows/settings.rs [hunk 2]
```diff
@@ -1353,6 +1394,354 @@ fn setup_button_section(
...
+            y: y_offset,
+        },
+        size: CGSize {
+            width: WINDOW_WIDTH - (MARGIN * 2.0),
+            height: LABEL_HEIGHT,
+        },
+    });
+    content_view.addSubview(&label);
+    std::mem::forget(label);
+    y_offset -= LABEL_HEIGHT + 5.0;
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/ui/windows/settings.rs]: Fixed in commit 855346c.

File: src/ui/windows/settings.rs [hunk 3]
```diff
@@ -1353,6 +1420,360 @@ fn setup_button_section(
...
+            let mut config = get_current_config();
+
+            // Add to the list
+            let mut keys = config.allowed_keys.unwrap_or_default();
+
+            // Check for duplicates
+            if keys.contains(&trimmed.to_string()) {
+                set_validation_label(&settings::ADD_KEY_VALIDATION, false, "Key already in list");
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/ui/windows/settings.rs]: Fixed in commit 121ecf2. Duplicate detection now uses case-insensitive comparison (`to_lowercase()`) to prevent semantic duplicates like "Cmd+A" and "cmd+a" from being added.

File: src/ui/windows/settings.rs [hunk 4]
```diff
@@ -1353,6 +1420,360 @@ fn setup_button_section(
...
+
+            // Clear the input field
+            unsafe {
+                with_ptr_void::<NSTextField, _>(&settings::ADD_KEY_FIELD, |field| {
+                    field.setStringValue(ns_string!(""));
+                });
+            }
+
+            // Update config in memory (not saved until user clicks Save)
+            set_current_config(config);
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/ui/windows/settings.rs]: See reply on comment at line 330 - this is part of the same Cancel behavior inconsistency. Will track as a follow-up issue.

File: src/ui/windows/settings.rs [hunk 5]
```diff
@@ -1353,6 +1420,360 @@ fn setup_button_section(
...
+
+    // Update config in memory
+    set_current_config(config);
+
+    set_validation_label(
+        &settings::ADD_KEY_VALIDATION,
+        true,
+        "✓ Cleared (click Save to persist)",
+    );
+}
```
[the remark this answers is not in the document]
Comment (taearls) [reply src/ui/windows/settings.rs]: See reply on comment at line 330 - this is part of the same Cancel behavior inconsistency. Will track as a follow-up issue.

**Review by taearls — COMMENTED** [review]
## Summary

This PR implements a comprehensive key allowlist feature (Issue #64) that allows users to specify keyboard shortcuts that pass through the shield overlay. This is a well-structured, sizeable feature addition.

**Changed files**: 10 files, +988 additions, -30 deletions
**Impact areas**: Input handling, Config system, Settings UI, Shield behavior
**Review depth**: Full validation

### Quality Checks Results

- ✅ Formatting: `cargo fmt --check` - Pass
- ✅ Linting: `cargo clippy -- -D warnings` - Pass (no warnings)
- ✅ Tests: `cargo test --lib` - Pass (193 tests, 0 failed)
- ✅ Build: `cargo build --release` - Pass

### Code Review Findings

#### ✅ Positive Observations

1. **Excellent code organization** - New `allowed_keys.rs` module is cleanly separated with proper public API
   - `src/input/allowed_keys.rs` follows single-responsibility principle
   - Clear documentation with module-level and function-level doc comments

2. **Thread-safe design** - Uses `RwLock` for global state with proper poisoned lock recovery
   - `src/input/allowed_keys.rs:93-109` handles poisoned mutex gracefully
   - Consistent pattern across all accessor functions

3. **Comprehensive test coverage** - 13 new unit tests covering:
   - Parsing with/without modifiers
   - Invalid input handling
   - Global state management
   - Event matching logic
   - Edge cases for exact modifier matching

4. **Good code reuse** - Leverages existing `ExitKey` parser infrastructure
   - `src/input/allowed_keys.rs:29-41` reuses parser, relaxing modifier requirement
   - Shared `ERR_NO_MODIFIER` constant avoids fragile string matching (fixed in recent commits)

5. **Proper error handling** - Returns descriptive errors for invalid keys
   - `src/input/allowed_keys.rs:62-63` provides helpful error messages listing valid keys

6. **Clean event tap integration** - Allowed keys check is minimal and efficient
   - `src/platform/event_tap.rs:83-87` simple check with early return

#### 🟡 Known Issue (Tracked)

**Cancel behavior for allowed_keys** - Already acknowledged and tracked as Issue #80
- `src/ui/windows/settings.rs:327-330` - Updates in-memory config immediately
- Other settings only update UI until Save - this inconsistency is documented

#### 🔵 Minor Suggestions

1. **Test isolation** - Tests modify global `ALLOWED_KEYS` state which can cause flaky tests when run in parallel
   - Consider adding `clear_allowed_keys()` in test setup/teardown
   - File: `src/input/allowed_keys.rs:210-265`

2. **Consider `#[must_use]` for parse functions**
   - `AllowedKey::parse()` and `parse_and_set_allowed_keys()` return `Result` that should be checked
   - File: `src/input/allowed_keys.rs:29`, `src/input/allowed_keys.rs:112`

### Testing Analysis

- **Coverage**: New/changed code is well tested with 13 new unit tests + 6 config tests
- **Test levels**: Appropriate unit tests for parsing, matching, and state management
- **Edge cases**: Good coverage for invalid input, exact modifiers, and poisoned mutex
- **Test quality**: Clear assertions, tests match implementation behavior

### Architecture & Patterns Compliance

- **Follows existing patterns**: Delegate pattern for text field validation matches `ExitKeyFieldDelegate` and `TimerFieldDelegate`
- **Consistent with codebase**: Uses same `AtomicPtr<c_void>` pattern for UI state management
- **Separation of concerns**: Core logic in `allowed_keys.rs`, UI in `settings.rs`, integration in `event_tap.rs`

### Security Review

- ✅ No exposed secrets in code
- ✅ Input validation present for key parsing
- ✅ No injection vulnerabilities
- ✅ Thread-safe state management with RwLock
- ✅ Graceful poisoned lock recovery prevents panics

### Performance Review

- **Algorithmic efficiency**: O(n) key matching is appropriate for small allowlist
- **Lock contention**: Uses `RwLock` allowing concurrent reads during event checking
- **Hot path**: `is_key_allowed()` in event callback is minimal (read lock + linear scan)

### Documentation Review

- ✅ Module-level documentation in `allowed_keys.rs`
- ✅ Function-level doc comments for public API
- ✅ ROADMAP.md updated with completion status
- ✅ Inline comments explaining complex logic

### Approval Status

✅ **READY TO MERGE** - This is a well-implemented feature with comprehensive tests, proper error handling, and clean integration. The known Cancel behavior issue is already tracked in Issue #80.

**Recent fixes applied**:
- Fixed `create_label` compilation errors
- Added `ERR_NO_MODIFIER` constant for maintainable error detection
- Fixed case-insensitive duplicate detection
- Fixed settings menu re-enablement after window close
- Fixed visual overlap of allowed keys section with opacity slider

---

**Review completed using**: `cargo fmt --check`, `cargo clippy -- -D warnings`, `cargo test --lib`, `cargo build --release`
**Reviewed by**: Claude Code Agent
